### PR TITLE
[Product Refactor] Primary Taxon

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Part of #9069 

Migrates `primary_taxon` from Product to Variant.

#### Review Note

I hit a bit of an issue described here: https://openfoodnetwork.slack.com/archives/C01T75H6G0Z/p1691581195059389 

I've added a temporary solution which resolves it (see last two commits) by applying the filtering products by taxon via the first taxon of the first variant and it avoids having to remove the ordering feature for now. I think we can come back to this later.

#### What should we test?

Setting the product category when creating a new product or editing a variant, filtering and ordering by taxon in the shopfront.

Displaying the taxon (product category) in the products and inventory report.

#### Release notes

Changelog Category: User facing changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Review by mkllnk — APPROVED** [review]
Nice work. Looks like we may need to rethink the shopfront ordering at some point.

Another couple of thoughts:

* A column ignore would prepare us for the column removal and test that Rails really doesn't try to access it.
* Resolving style issues or at least updating the Rubocop todo helps others keeping the code quality high.

File: app/controllers/api/v0/order_cycles_controller.rb
Comment (dacook) [inline app/controllers/api/v0/order_cycles_controller.rb:73]: It would be good to know why this filtering is now unnecessary.

File: app/services/products_renderer.rb
Comment (dacook) [inline app/services/products_renderer.rb:57]: I'm wondering what this method is needed for? It looks like the old relation used just as much arel as the new one.

File: app/services/order_cycle_distributed_products.rb
Comment (dacook) [inline app/services/order_cycle_distributed_products.rb:18]: I think we can try using left_joins to make this slightly shorter:
```suggestion
      left_joins("(SELECT DISTINCT ON(product_id) id, product_id, primary_taxon_id FROM spree_variants) first_variant ON spree_products.id = first_variant.product_id").
```
We'll also need to fix the linter error.

I haven't got any better ideas right now sorry..

**Review by dacook — CHANGES_REQUESTED** [review]
This looks good. I have a few questions and think we might be able to simplify the factory.

But I think the only essential change is to fix the linter errors. I'll have a go at that now.

File: db/migrate/20230807122052_migrate_product_taxons.rb
Comment (dacook) [reply db/migrate/20230807122052_migrate_product_taxons.rb:11]: I'm not sure why, but after checking out locally and running all migrations, I still get this error when running specs:

```
PG::NotNullViolation: ERROR:  null value in column "primary_taxon_id" of relation "spree_products" violates not-null constraint
```

File: app/services/order_cycle_distributed_products.rb
Comment (dacook) [reply app/services/order_cycle_distributed_products.rb]: I couldn't get `left_joins` to work, so have kept it as-is (with added line-breaks for code formatting).

**Review by dacook — APPROVED** [review]

**Comment by drummer83** [comment]
Hi @Matt-Yorkley,
There's a conflict here.
Moving to In Dev for resolution.
Thanks!

**Review by mkllnk — APPROVED** [review]
Rebased and conflicts resolved.

**Comment by mkllnk** [comment]
I resolved the conflicts but there are quite a few specs failing now. Back to In Dev for @Matt-Yorkley. We need to review these pull requests quicker to avoid conflicts. Otherwise it will be stuck in the review-test-rebase loop forever.

File: app/services/products_renderer.rb
Comment (Matt-Yorkley) [reply app/services/products_renderer.rb:57]: Pagy wasn't correctly handling the query here, leading to some failing specs due to pagination errors. Something about the join being added was confusing it. This `pagy_arel` method just makes pagy use some slightly more nuanced query parsing, which in this case meant it could handle the join correctly.

File: app/controllers/api/v0/order_cycles_controller.rb
Comment (Matt-Yorkley) [reply app/controllers/api/v0/order_cycles_controller.rb:73]: There's a whitelisting of params that are used by Ransack which takes place automatically, and the allowed attributes and scopes are defined on models, eg:

https://github.com/openfoodfoundation/openfoodnetwork/blob/a096c56658bd9dd473898b3970bca8acc16b389e/app/models/spree/product.rb#L36-L38

So this big chunk of messy code is redundant; it's doing something that's already being done elsewhere.

File: spec/models/spree/product_spec.rb
Comment (Matt-Yorkley) [reply spec/models/spree/product_spec.rb]: I added a commit :+1:

File: app/controllers/api/v0/order_cycles_controller.rb
Comment (dacook) [reply app/controllers/api/v0/order_cycles_controller.rb:73]: Hooray!

**Review by dacook — APPROVED** [review]
Great, I reviewed the last commits, all looks good to me

**Comment by dacook** [comment]
Ready to test with no conflicts!

**Comment by dacook** [comment]
As it's user-facing change, it would be good to provide a little more detail in the PR name. I'm not sure exactly how it will affect users, but I guess they will see the category selection when editing Variants instead of Products? Does the tester have a suggestion? 🥺

**Comment by drummer83** [comment]
Hi @Matt-Yorkley,
Thanks for continuing your work on these network 2.0 preparations. :muscle: 

I have tested the following.

**Super admin:**
- Create new taxons :heavy_check_mark: 
- Edit taxons :heavy_check_mark: 
- Delete taxons :heavy_check_mark: (unless #9387)

**Admin:**
- Create new product :heavy_check_mark: 
- Create new variant :heavy_check_mark: (price is not checked to be a number anymore - it's set to 0.00)
- Duplicate product :heavy_check_mark: (on demand setting is not duplicated)
- Edit variant :heavy_check_mark: 
- Delete variant :heavy_check_mark: 
- Import products :heavy_check_mark: (working fine for the first import, second import fails due to #11492, but no error for product category is shown - so I assume it would pass)
- Define sorting of products by category in shop front :heavy_check_mark: 
- Products & Inventory Reports :heavy_check_mark: 
- Filter in product list :x:
  - All variants are shown if one variant meets the product category (this may be part of your compromise and we could live with it).
  - If there are n variants of the same product with the same product category and the filter is set to that category, then the product (including all variants) is shown n times in the product list.
image

**Shopper:**
- Filter for categories :heavy_check_mark: 
  - The filter ignores deleted variants properly.
- Sorted products by category :x: 
  - Deleted variants are taken into account when defining the sorting in the shop. This can lead to incorrect results.
image
image

### Conclusion
I am not sure how edgy the two failing test cases are. They both are regressions, I think. After our discussion on Tuesday about the trade off between quality and invested time I don't know if we want to merge, create issues and maybe fix later or if we want fix this before merging. So I'm adding the feedback needed label and ask @openfoodfoundation/train-drivers-product-owners for a decision.

Thanks anyway for this great effort! :pray:

**Comment by RachL** [comment]
Thx @drummer83 for these thorough testing notes!

I couldn't make it on Tuesday so I'm not sure I have all the info. Are those last bits tricky to fix @Matt-Yorkley ? I'm thinking we could merge and fix later: it will take a while for users to see they can change the category per variant and even in that case, scenarios where they will use it are rare - but I could be wrong.

**Comment by drummer83** [comment]
I thought it's a good idea to test deeply due to the regression we currently have with the product import and tax categories. Didn't want that to happen again. :wink:

I agree that use cases for different categories for variants of the same product are rare and will probably not be used for some time.
However, as soon as users filter the product list for categories, they will see the multiple entries. This is "not nice" but still everything is working. No strong opinion on my side but regarding the fixing of the issues these multiple entries should be the priority.

**Comment by RachL** [comment]
Ah so the multiple entries are happening all the time, not only when variants of the same product have different categories. In that case yes I agree it should be fixed in priority.

File: spec/reflexes/products_reflex_spec.rb
Comment (dacook) [inline spec/reflexes/products_reflex_spec.rb:47]: Nice pickup!

File: app/services/order_cycle_distributed_products.rb
Comment (dacook) [inline app/services/order_cycle_distributed_products.rb:17]: I had to read it a couple of times to work out what it meant ;)
```suggestion
  # enterprise can display product sorted by category in a custom order on their shopfront.
```

**Review by dacook — COMMENTED** [review]
Great, that didn't look so bad! Looks like it was a good introduction to the topic.

I just noticed a typo which I'll update next.

**Review by dacook — APPROVED** [review]

File: app/services/order_cycle_distributed_products.rb
Comment (mkllnk) [inline app/services/order_cycle_distributed_products.rb:29]: I think that this is the query that should ignore deleted variants. From Konrad's testing:

> Deleted variants are taken into account when defining the sorting in the shop. This can lead to incorrect results.

This is an edge case at the moment though and I'm happy to go ahead with this PR and fix it later.

File: app/services/order_cycle_distributed_products.rb
Comment (mkllnk) [inline app/services/order_cycle_distributed_products.rb:26]: I was wondering how we select only the first variant. After a while I realised that `DISTINCT ON(product_id)` ensures that we select only one variant per product but I guess that it could be a random one.

**Review by mkllnk — APPROVED** [review]
Great! It will be good to have this monster PR out of the way.

File: app/services/order_cycle_distributed_products.rb
Comment (rioug) [reply app/services/order_cycle_distributed_products.rb:26]: We could add an `ORDER BY created_at` on the sub query and ignore deleted ones so we always pick up the first created one ?

Comment (mkllnk) [reply app/services/order_cycle_distributed_products.rb:26]: I don't think it matters which one we select as long as we ignore deleted ones.

**Comment by filipefurtad0** [comment]
Hi all,

I've tried different scenarios, but still can observe both issues @drummer83 spotted earlier:

- filtering by product category displays the product n times, for n variants with category x

image

- ordering in shopfront considers deleted variants

image

**Comment by filipefurtad0** [comment]
Thinking about this again, I thought it would be a good idea to re-stage the PR, and deleting the migrations to make sure they are ran again. Deleting worked, as usual, but staging failed:

https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/65e9f0b408c6fa00070d0cd5?event_id=65e9f0b400dd8108939d0000&i=sk&m=nw

Thoughts?

**Comment by filipefurtad0** [comment]
Ok, so after staging the first bug (1) is still reproducible:

image

The second (2) however, is fixed:

Before deleting the variant responsible for bringing the product (`cat-o-mat`) to the highest rank:

image

After deleting it, we can see that the product (`cat-o-mat`) lowers rank, as expected:

image

From the [discussion above](https://github.com/openfoodfoundation/openfoodnetwork/pull/11369#issuecomment-1712470446) it does seem that this one is a blocker... 

It's too bad too hold this one back. On the other hand, it might get messy when hubs have too many products. I guess we move this one back to in dev. Pinging @openfoodfoundation/train-drivers-product-owners and @rioug for awareness.

**Review by mkllnk — APPROVED** [review]

**Review by dacook — APPROVED** [review]

**Comment by drummer83** [comment]
Hi @rioug,
Sorry, there's a conflict here now. Could you please have a look?
Thanks!

**Comment by rioug** [comment]
@drummer83 This PR touches a lot of code so it's very easy to get a conflict, could you let me know when you'll have time to test this ? I want to fix it at the last minute to avoid running into a conflict again.

**Comment by filipefurtad0** [comment]
Hey @rioug ,

There were some issues staging the PR, it seems the first attempt fails to fully deploy; this happened for both:

**(1) staging-FR**

Here, staging was eventually sucessful on the second run, but I noticed that some products are gone, from the products page:

image

`cat-o-mat` was present, just before staging, and like in [previous tests](https://github.com/openfoodfoundation/openfoodnetwork/pull/11369#issuecomment-1985537825).

I'll stage master to make sure these products appear again, and this is indeed introduced by this PR.

**(2) staging-UK**

I thought this could be an issue with the need of re-running the migration (for which, we'd have to delete a pre-existing version). But I could not find any of the migrations from this PR:

```
openfoodnetwork=> select * from schema_migrations WHERE version LIKE '%2023080%';
    version     
----------------
 20230807145022
 20230809172206
 20230809194304
 20230809195519
 20230809201542
(5 rows)

```

I'll still test staging master before - since the PR was recently rebase, maybe a migration which is in master could be needed? I'm not sure, can't hurt trying.

**Comment by rioug** [comment]
> However, I'm not sure about the deployment issues for staging-UK.. Master stages just fine, so I find this a bit concerning. What do you think? This is the only blocker for merging.

It looks like the DB on staging-UK already has `primary_taxon_id` added on `spree_variants` table, which is weird. It shouldn't be an issue on production though.

**Comment by rioug** [comment]
> However, I'm not sure about the deployment issues for staging-UK.. Master stages just fine, so I find this a bit concerning. What do you think? This is the only blocker for merging.

The staging-UK DB somehow got out of sync, it already has the column the migration is trying to add. 
I tested the migration on a copy of au prod DB and it's fine, I going to merge.

**Pull request merged**
